### PR TITLE
Introduce explicit way of halting callback chains by throwing :abort. Deprecate current implicit behavior of halting callback chains by returning `false` in apps ported to Rails 5.0. Completely remove that behavior in brand new Rails 5.0 apps.

### PR DESCRIPTION
**Pull request by claudiob**
Update (2015-01-02): a gist with the suggested release notes to add to Rails 5.0 after this commit is available at https://gist.github.com/claudiob/614c59409fb7d11f2931

---

Stems from discussion with @dhh at https://groups.google.com/forum/#!topic/rubyonrails-core/mhD4T90g0G4

@dhh – I created this work-in-progress PR to continue the conversation with some code.
Could you tell me if this is what you intended by using `:throw` with a symbol?

I have quite clear how to add the `throw` to the code in ActiveJob, but not as
much where to `catch` it in ActiveSupport. For now I have this code:

``` ruby
def run_callbacks(kind, &block)
  catch(:abort_job)  do
    send "run_#{kind}_callbacks", &block
  end
end
```

but maybe the `catch` is better located somewhere at a deeper level of the callback stack.

@dhh Thoughts? Once I have a clearer idea on how to proceed, I can complete
this PR with tests, documentations, etc. Thanks!

### CODE REVIEW

**Comment by dhh** [comment]
Great work on this. Is there any additional unresolved issues or are we good to merge?

**Comment by sgrif** [comment]
I think we just need a squash, rebase, and CI run

**Comment by claudiob** [comment]
@sgrif I just rebased and CI is running. 

Are you sure you want a squash :jack_o_lantern: ? I tried to keep each commit valid on its own.

**Comment by sgrif** [comment]
No, that's fine.

**Comment by dhh** [comment]
Great. Lovely work, Claudio. Very diligent.

File: activerecord/lib/active_record/autosave_association.rb [hunk 1]
```diff
@@ -200,13 +200,13 @@ def add_autosave_association_callbacks(reflection)
...
           else
-            define_non_cyclic_method(save_method) { save_belongs_to_association(reflection) }
+            define_non_cyclic_method(save_method) { throw(:abort) if save_belongs_to_association(reflection) == false }
             before_save save_method
           end
 
           if reflection.validate? && !method_defined?(validation_method)
             method = (collection ? :validate_collection_association : :validate_single_association)
-            define_non_cyclic_method(validation_method) { send(method, reflection) }
+            define_non_cyclic_method(validation_method) { send(method, reflection) || nil }
```
Comment (sgrif) [inline activerecord/lib/active_record/autosave_association.rb]: Why are we converting false to nil here?

File: railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Be sure to restart your server when you modify this file.
```
Comment (sgrif) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb]: Do we need to mention this? This is true of all initializers.

Comment (claudiob) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb]: Yes, and that line is at the top of every initializer, so I just added it for consistence with:

* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/wrap_parameters.rb.tt#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/session_store.rb.tt#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/mime_types.rb#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/inflections.rb#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/filter_parameter_logging.rb#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/cookies_serializer.rb#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/backtrace_silencers.rb#L1
* https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/assets.rb.tt#L1

**Comment by claudiob** [comment]
@dhh After discussing with @sgrif and @matthewd, we might want to *delete* the initializer before we merge this PR. Let me explain and see if you agree too.

By itself, having a config option `Rails.application.config.active_support.halt_callback_chains_on_return_false` is a good idea: savvy developers and gem creators can decide whether they want to switch 100% to the new behavior (without even deprecation warnings) or temporarily keep the current behavior (with a deprecation warning).

However, adding an initializer `config/initializers/callback_terminator.rb` that sets the value of the config option to `false` might be dangerous, given the deprecation policy of Rails.

---

Here is an example: say I am the creator of a gem (Devise? Clearance?) which uses callbacks, and returns `false` to halt such callbacks. 

As soon as Rails 5 comes out, I want to release a new version that is compatible with Rails 5. I run my test suite and it still passes, albeit with deprecation warnings telling me that I *should* `throw(:abort)`, but `return false` works just fine. So I release the new version of the gem.

Now a user creates a brand new Rails app and adds my gem to the Gemfile. Because of the initializer, the gem will not work! All those callbacks returning `false` that only raised a deprecation warning in the test suite will stop working, and the user won't even know why.

Given this, it might be *safer* to drop the initializer from this PR, fix some of the Markdown documentation related to that, and have Rails 5.0 stick to the current deprecation cycle, then switch 100% to the new behavior in Rails 5.1.

Yes, Rails 5.1 is far away, but it might be better to get there safely without getting developers and gem maintainers angry. What do you think?

**Comment by dhh** [comment]
I don’t think we should extend the deprecation cycle to include all sorts of plugins and gems that may or may not be updated. In the case of Devise, we can solve the case by having the plugin check for the variable and its setting to determine how to halt.

**Comment by matthewd** [comment]
@claudiob let's just drop the initializer from this PR. That way, if we want to, we can add it separately, and argue the default behaviour in a separate PR, on its own.

If we do decide to force the change, we'll want a nice clean place we can point people to when they complain, anyway.

**Comment by dhh** [comment]
Dealing with plugins that haven’t been upgraded for new applications is not a compelling enough argument for me to push this out beyond 2015. We shouldn’t forget the pain and bugs the existing behavior is causing. Accidentally returning false is all too easy as it is. So I’m +1 on keeping the default-for-new-apps as part of this PR.

**Comment by sgrif** [comment]
The deprecation cycle is the same regardless. We can't just remove the old behavior in 5.0. We commit to a full version deprecation cycle for any breaking changes. Gem authors frequently can and do update their gem and release a new version, without having fixed every single deprecation warning, because we tell them it's safe to do so. By changing the default for new apps, we're making a breaking change.

This becomes even worse when you consider that it is unlikely for things to blow up when an app with the new config uses a gem that was relying on the old behavior. It's instead likely to introduce subtle bugs which are hard to detect unless you know exactly where every gem you use might define callbacks.

**Comment by matthewd** [comment]
@dhh the deprecation warning would inform you that you're accidentally returning false.

I vehemently disagree with the suggestion that there's a "but only the actual application, no gems allowed" exception in our published deprecation policy.

**Comment by dhh** [comment]
We’re not removing the old behavior. We’re saying that new apps started after the release of Rails 5.0 gets new behavior. Plugins that haven’t been updated to support Rails 5.0 shouldn’t be used with that new major version.

Existing apps get the standard deprecation cycle.

We’ve followed this pattern in the past for other default changes and I’m happy with the timeline that presents.

File: activerecord/lib/active_record/autosave_association.rb (hunk 1 above)
Comment (claudiob) [inline activerecord/lib/active_record/autosave_association.rb]: Whenever a callback returns `false`, the user gets a deprecation warning.

In the rest of the code, I have changed all internal uses of `return false` to `throw(:abort)`, so users will not get deprecation warnings for internal code.

However, in this specific case, there is no need to do that because the value returned by this method (whether `nil` or `false`) is not used anywhere. The method above is [invoked here](https://github.com/rails/rails/blob/8b07f18599854475a69c1447642dd3d80bea8df9/activemodel/lib/active_model/validations.rb#L394) inside `_run_validate_callbacks`:

```ruby
    def run_validations! #:nodoc:
      _run_validate_callbacks
      errors.empty?
    end
```

The only important thing is whether `errors` is empty or not.
Since it's not meaningful whether the callback returns `false` or `not`, I added the `|| nil`, so the user will not get a useless deprecation warning. 

---

The following is to show you how you would get deprecation warning when running tests if you didn't add the `|| nil`:

```bash
DEPRECATION WARNING: Returning `false` in a callback will not implicitly halt a callback chain in the next release of Rails. To explicitly halt a callback chain, please use `throw :abort` instead. (called from block in <class:TestAutosaveAssociationValidationsOnABelongsToAssociation> at test/cases/autosave_association_test.rb:1502)
/Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:617:in `display_deprecation_warning_for_false_terminator'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:607:in `block (2 levels) in default_terminator'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:604:in `catch'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:604:in `block in default_terminator'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:166:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:166:in `block in halting'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `block in call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `each'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:506:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:93:in `_run_callbacks'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:808:in `_run_validate_callbacks'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations.rb:397:in `run_validations!'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations/callbacks.rb:112:in `block in run_validations!'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:89:in `call'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:89:in `_run_callbacks'
  /Users/claudiob/code/rails/activesupport/lib/active_support/callbacks.rb:808:in `_run_validation_callbacks'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations/callbacks.rb:112:in `run_validations!'
  /Users/claudiob/code/rails/activemodel/lib/active_model/validations.rb:334:in `valid?'
  /Users/claudiob/code/rails/activerecord/lib/active_record/validations.rb:58:in `valid?'
  test/cases/autosave_association_test.rb:1502:in `block in <class:TestAutosaveAssociationValidationsOnABelongsToAssociation>'
```

**Comment by claudiob** [comment]
@dhh @tenderlove I tried to address the concerns about performance with a benchmark on the impact of this PR on ActiveRecord. The results are here: https://gist.github.com/claudiob/c45310439d4090de5c90.

**Comment by dhh** [comment]
Excellent. Are there any other blockers left or can we merge?

File: activerecord/lib/active_record/autosave_association.rb (hunk 1 above)
Comment (rafaelfranca) [inline activerecord/lib/active_record/autosave_association.rb]: This means we can remove the `|| nil` when we remove the deprecation warning. If that is really the case a TODO comment explaining this would be good.

File: activesupport/lib/active_support/i18n_railtie.rb [hunk 3]
```diff
@@ -55,7 +55,7 @@ def self.initialize_i18n(app)
 
       reloader = ActiveSupport::FileUpdateChecker.new(I18n.load_path.dup){ I18n.reload! }
       app.reloaders << reloader
-      ActionDispatch::Reloader.to_prepare { reloader.execute_if_updated }
+      ActionDispatch::Reloader.to_prepare { reloader.execute_if_updated.presence }
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/i18n_railtie.rb]: Why this need to call `presence`?

File: railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb (hunk 2 above)
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb]: The `rake rails:update` task should not generate this file.

File: activerecord/lib/active_record/autosave_association.rb (hunk 1 above)
Comment (claudiob) [inline activerecord/lib/active_record/autosave_association.rb]: @rafaelfranca Should I just add a comment line like this?

```ruby
# TODO: remove the following +|| nil+ as soon as returning +false+ in a callback 
# does not display a deprecation warning 
define_non_cyclic_method(validation_method) { send(method, reflection) || nil }
```

Comment (rafaelfranca) [inline activerecord/lib/active_record/autosave_association.rb]: Yes, but since it will not generate rdoc we can use markdown.

File: activesupport/lib/active_support/i18n_railtie.rb (hunk 3 above)
Comment (claudiob) [inline activesupport/lib/active_support/i18n_railtie.rb]: Before this PR, returning `false` in an ActionDispatch callback did not have any side effect.

After this PR, returning `false` in an ActionDispatch callback halts the chain and shows a deprecation warning.

The Rails codebase uses two ActionDispatch callbacks [defined here](https://github.com/rails/rails/blob/8a1c3476a62005e3cb7cfe440da71b94fab8aa5b/actionpack/lib/action_dispatch/middleware/reloader.rb#L32L51), namely `ActionDispatch::Reloader#to_prepare` and `#to_cleanup`.

In the code above, I added `presence` to prevent the body of the `#to_prepare` callback from returning `false` and, therefore, halting the callback. In fact the `execute_if_updated` method above explicitly returns `false` sometimes ([source code](https://github.com/rails/rails/blob/8a1c3476a62005e3cb7cfe440da71b94fab8aa5b/activesupport/lib/active_support/file_update_checker.rb#L82L89)):

```ruby
def execute_if_updated
  if updated?
    execute
    true
  else
    false
  end
end
```

---

Looking back at the code now, I realize that a better alternative to the `.presence` above might be to explicitly declare a custom terminator for  `ActionDispatch::Reloader#to_prepare` and `#to_cleanup`, overriding for these specific case the default terminator, so they never halt the chain, not even when `false` is returned or `:abort` is thrown. 

In other words, replacing [these two lines](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/middleware/reloader.rb#L32L33):

```ruby
define_callbacks :prepare
define_callbacks :cleanup
```

with something like:

```ruby
define_callbacks :prepare, terminator: ->(target, result) { false }
define_callbacks :cleanup, terminator: ->(target, result) { false }
```

@rafaelfranca Would you rather go this direction?

File: railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb (hunk 2 above)
Comment (claudiob) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb]: What do I need to change to prevent that from happening?

File: activesupport/lib/active_support/i18n_railtie.rb (hunk 3 above)
Comment (rafaelfranca) [inline activesupport/lib/active_support/i18n_railtie.rb]: Yes, I think the explicit terminator is a better direction. I'm pretty sure we are expecting prepare callbacks to stop if returning false.

File: railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb (hunk 2 above)
Comment (rafaelfranca) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb]: Just remove the file on this method https://github.com/rails/rails/blob/7c6861b0bc85077ed1f1dcb4a919de0f1b8d490d/railties/lib/rails/generators/rails/app/app_generator.rb#L89 after the `config` call.

File: activesupport/lib/active_support/i18n_railtie.rb (hunk 3 above)
Comment (claudiob) [inline activesupport/lib/active_support/i18n_railtie.rb]: For completeness, this is how tests would fail if we let ActionDispatch callbacks halt the chain when returning `false`:

```bash
~/code/rails/railties$ ruby -w -Itest:lib:/Users/claudiob/code/rails/railties/../activesupport/lib:/Users/claudiob/code/rails/railties/../actionpack/lib:/Users/claudiob/code/rails/railties/../activemodel/lib test/application/console_test.rb
/Users/claudiob/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/bundler-1.7.9/lib/bundler/source/rubygems.rb:191: warning: shadowing outer local variable - spec
Run options: --seed 38479

# Running:

.DEPRECATION WARNING: Returning `false` in a callback will not implicitly halt a callback chain in the next release of Rails. To explicitly halt a callback chain, please use `throw :abort` instead. (called from block in test_reload_should_fire_preparation_and_cleanup_callbacks at test/application/console_test.rb:48)
F...DEPRECATION WARNING: Returning `false` in a callback will not implicitly halt a callback chain in the next release of Rails. To explicitly halt a callback chain, please use `throw :abort` instead. (called from block in test_reload_should_reload_constants at test/application/console_test.rb:72)
.

Finished in 7.202340s, 0.8331 runs/s, 4.0265 assertions/s.

  1) Failure:
ConsoleTest#test_reload_should_fire_preparation_and_cleanup_callbacks [test/application/console_test.rb:52]:
Expected: 3
  Actual: 2

6 runs, 29 assertions, 1 failures, 0 errors, 0 skips
```

Comment (rafaelfranca) [inline activesupport/lib/active_support/i18n_railtie.rb]: If we can't halt of maybe it is better to use the `|| nil` again.

Comment (claudiob) [inline activesupport/lib/active_support/i18n_railtie.rb]: > I'm pretty sure we are expecting prepare callbacks to stop if returning false. 

I didn't understand this sentence. In the current Rails code, ActionDispatch callbacks do not halt the chain when returning `false`. That setting is currently only present in [ActiveModel](https://github.com/rails/rails/pull/17227/files#diff-174733f2db65ef1bc53e3222d6ac0e61L20), in [ActiveModel::Validations](https://github.com/rails/rails/pull/17227/files#diff-0cd7b1cc8e754e9ab96c72b45770767dL26), and in [ActiveRecord](https://github.com/rails/rails/pull/17227/files#diff-174733f2db65ef1bc53e3222d6ac0e61L20).

Since currently we **don't** expect ActionDispatch callbacks to halt upon returning `false`, and the new default terminator **will** halt upon returning `false` and display a warning (unless `config.active_support.halt_callback_chains_on_return_false` is set to `false`), then I would add the following *custom* terminators for `#to_prepare` and `#to_cleanup`, so those callbacks will keep on **not halting** upon returning `false`:

```ruby
define_callbacks :prepare, terminator: ->(_, result_lambda) { result_lambda.call if result_lambda.is_a?(Proc); false }
define_callbacks :cleanup, terminator: ->(_, result_lambda) { result_lambda.call if result_lambda.is_a?(Proc); false }
```

Makes sense?

File: activerecord/lib/active_record/autosave_association.rb [hunk 4]
```diff
@@ -200,13 +200,19 @@ def add_autosave_association_callbacks(reflection)
...
 
           if reflection.validate? && !method_defined?(validation_method)
             method = (collection ? :validate_collection_association : :validate_single_association)
-            define_non_cyclic_method(validation_method) { send(method, reflection) }
+            define_non_cyclic_method(validation_method) do
+              send(method, reflection)
+              # TODO: remove the following line as soon as the return value of
+              # callbacks is ignored, that is, returning `false` does not
+              # display a deprecation warning or halts the callback chain.
+              true
```
Comment (claudiob) [inline activerecord/lib/active_record/autosave_association.rb]: @rafaelfranca I changed this as you suggested, so it has the TODO statement and simply returns `true`.

File: activesupport/lib/active_support/i18n_railtie.rb [hunk 5]
```diff
@@ -55,7 +55,13 @@ def self.initialize_i18n(app)
 
       reloader = ActiveSupport::FileUpdateChecker.new(I18n.load_path.dup){ I18n.reload! }
       app.reloaders << reloader
-      ActionDispatch::Reloader.to_prepare { reloader.execute_if_updated }
+      ActionDispatch::Reloader.to_prepare do
+        reloader.execute_if_updated
+        # TODO: remove the following line as soon as the return value of
+        # callbacks is ignored, that is, returning `false` does not
+        # display a deprecation warning or halts the callback chain.
+        true
```
Comment (claudiob) [inline activesupport/lib/active_support/i18n_railtie.rb]: @rafaelfranca I also changed this as you suggested, so it has the TODO statement and simply returns true.

File: railties/lib/rails/generators/rails/app/app_generator.rb [hunk 6]
```diff
@@ -88,12 +88,19 @@ def config
 
     def config_when_updating
       cookie_serializer_config_exist = File.exist?('config/initializers/cookies_serializer.rb')
+      callback_terminator_config_exist = File.exist?('config/initializers/callback_terminator.rb')
 
       config
 
+      unless callback_terminator_config_exist
+        remove_file 'config/initializers/callback_terminator.rb'
+      end
```
Comment (claudiob) [inline railties/lib/rails/generators/rails/app/app_generator.rb]: @rafaelfranca I have removed the initializer as you suggested

File: railties/test/generators/app_generator_test.rb [hunk 7]
```diff
@@ -160,6 +160,38 @@ def test_rails_update_keep_the_cookie_serializer_if_it_is_already_configured
...
+    Rails.application.config.root = app_root
+    Rails.application.class.stubs(:name).returns("Myapp")
+    Rails.application.stubs(:is_a?).returns(Rails::Application)
+
+    generator = Rails::Generators::AppGenerator.new ["rails"], { with_dispatchers: true }, destination_root: app_root, shell: @shell
+    generator.send(:app_const)
+    quietly { generator.send(:update_config_files) }
+    assert_file "#{app_root}/config/initializers/callback_terminator.rb"
+  end
+
```
Comment (claudiob) [inline railties/test/generators/app_generator_test.rb]: @rafaelfranca and I have added two tests here to prove that running `rake rails:update` on:

* an app *without* the initializer does not create the initializer
* an app *with* the initalizer already present for any reason does not touch the initializer

File: activemodel/CHANGELOG.md [hunk 8]
```diff
@@ -1 +1,12 @@
+*   Change the way in which callback chains can be halted.
+
+    The preferred method to halt a callback chain from now on is to explicitly
+    `throw(:abort)`.
+    In the past, returning `false` in an ActiveModel or ActiveModel::Validations
+    `before_` callback had the side effect of halting the callback chain.
+    This is not recommended anymore and, depending on the value of the
+    `config.active_support.halt_callback_chains_on_return_false` option, will
+    either not work at all or display a deprecation warning.
```
Comment (claudiob) [inline activemodel/CHANGELOG.md]: @matthewd I have updated the CHANGELOG files to be "stronger" than just saying "deprecated" behavior.
What do you think? (Please check the other CHANGELOG.md files too)

File: railties/test/generators/app_generator_test.rb (hunk 7 above)
Comment (rafaelfranca) [inline railties/test/generators/app_generator_test.rb]: :+1:

**Comment by claudiob** [comment]
:scream: :bowtie: :v: :clap:

And now… on with the next step. @dhh said:

> Ultimately, I think you should be able to halt the after_* chain as well. 
> But we don’t need to deal with it in the same PR.

This also related to #8479. What's the best plan to move forward with this? Is this something still wanted?

**Comment by dhh** [comment]
Bad ass, Claudio. You ran the gauntlet! :)

I still think halting the callback chain on after_ makes plenty of sense to me. Let’s just start a new PR for that.

**Comment by claudiob** [comment]
I created a presentation related to this PR for the L.A. Ruby meetup. 
If anyone's interested, here's the link: https://speakerdeck.com/claudiob/callbacks-in-rails-5

File: activesupport/lib/active_support/railtie.rb [hunk 9]
```diff
@@ -13,6 +13,13 @@ class Railtie < Rails::Railtie # :nodoc:
       end
     end
 
+    initializer "active_support.halt_callback_chains_on_return_false", after: :load_config_initializers do |app|
```
Comment (rafaelfranca) [inline activesupport/lib/active_support/railtie.rb]: @claudiob do you remember why we are running this after `load_config_initializers`? This is making `load_config_initializers` run way early in the initializers order.

Comment (claudiob) [inline activesupport/lib/active_support/railtie.rb]: @rafaelfranca Brand new Rails 5 come [with an initializer file called `config/initializers/callback_terminator.rb`](https://github.com/rails/rails/blob/master/railties/lib/rails/generators/rails/app/templates/config/initializers/callback_terminator.rb) which sets the value for `app.config.active_support.halt_callback_chains_on_return_false`.

In this code, we check if that value has been set and eventually assign it to `ActiveSupport::Callbacks::CallbackChain.halt_and_display_warning_on_return_false`.

Therefore, it made sense to me to run this method **after** the `config/initializers/callback_terminator.rb` has been called.

What kind of issues is this causing?

Comment (matthewd) [inline activesupport/lib/active_support/railtie.rb]: @rafaelfranca this sounds like a bug^W possibly unintended behaviour I've run across before. After you define an `initializer(before/after: ..)`, subsequent `initializer` definitions with no special ordering requirements (like the three below) will be inserted immediately after that one, instead of "falling" to their more natural position.

When I hit it, I just worked around it without investigating whether it was deliberate. Maybe it's not? Certainly my default expectation is that adding a definition like this should be a simple insertion, and have no effect on the order of the others.

Comment (rafaelfranca) [inline activesupport/lib/active_support/railtie.rb]: @claudiob this caused a subtle change in our initializing order https://github.com/troter/rails/commit/50f51b0ded5958c8104cd739a4220cc82d822df8.

@matthewd yeah. I had the same suspicion, but for what I could understand this is how our TSort implementation is working. I'll investigate better today, but I fell that if we change this we may have more problems.

Comment (rafaelfranca) [inline activesupport/lib/active_support/railtie.rb]: Fixed at 0a120a818d413c64ff9867125f0b03788fc306f8

**Comment by claudiob** [comment]
I created a presentation related to this PR for RailsConf 2015
If anyone's interested, here's the link: https://speakerdeck.com/claudiob/better-callbacks-in-rails-5

**Comment by claudiob** [comment]
==> https://github.com/rails/rails/issues/20612

**Comment by denniscollective** [comment]
Is there anyway the scope of this could be increased to allow you to throw in controller methods that are run as guard clauses rather than before_filters to abort the rest of the action.

It would be really nice for avoiding DoubleRenderErrors.

Trivial Case:

```ruby
class ThingsController < ApplicationController
def index
ensure_authorized
end

protected

def ensure_authorized
redirect_to root_url
throw(:abort)
end


end

```

I will gladly submit a pull request if it would be accepted, but don't want to waste time.

/cc @claudiob @dhh

**Comment by claudiob** [comment]
@denniscollective I think what you suggest would be overwhelming, meaning that we would have to wrap every action inside a `catch(:abort)`, and we would let every method (not just "guard clauses")  `throw(:abort)` and expect the action to be interrupted.

In the example above, what would you like to happen if the user is not authorized? 
If you have something in mind, you can explicitly write it as:

```ruby
def index
  if authorized
    # .. render view or any behavior that you desire
  else
    # .. redirect_to with a flash or any behavior that you desire
  end
end
```

If you really want the behavior provided by `throw(:abort)`, you can instead move `ensure_authorized` from inside the `index` method into `before_action :ensure_authorized, only: :index`.

**Comment by yamanaltereh** [comment]
@claudiob I use throw(:abort), but actually in `save!` it should return false not raise exception!

**Comment by claudiob** [comment]
@yamanaltereh Hello! Can you provide a little more context or a snippet of code?

If you found a bug, a [bug report](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report) would also be very useful. Thanks!

**Comment by yamanaltereh** [comment]
@claudiob Thank you, I'm not sure if it's a bug or not!
but after using `throw(:abort)` to terminate the callback, the `records.save!` (with "!") causes exception!, it should return true/false, right?

My code like this:
```ruby
class User < ApplicationRecord

  before_save :can_change_status, if: :status_changed?

  private

  def can_change_status?
    # ...
    can = false
    unless can
      self.errors.add(:user, '...')
      throw(:abort)
    end
  end
end

user = User.last
user.status = 'new_status'
user.save! # => causes exception, it should return true/false

```
